massa-ledger/src/cursor.rs

```diff
@@ -0,0 +1,132 @@
+use massa_hash::{Hash, HashDeserializer};
+use massa_models::{address::AddressDeserializer, Address};
+use massa_serialization::{Deserializer, SerializeError, Serializer};
+use nom::{sequence::tuple, IResult};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LedgerCursorStep {
+    Start,
+    Balance,
+    Bytecode,
+    Datastore(Option<Hash>),
+    Finish,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct LedgerCursorStepSerializer;
+
+impl LedgerCursorStepSerializer {
+    fn _new() -> Self {
+        Self
+    }
+}
+
+impl Serializer<LedgerCursorStep> for LedgerCursorStepSerializer {
+    fn serialize(&self, value: &LedgerCursorStep) -> Result<Vec<u8>, SerializeError> {
+        match value {
+            LedgerCursorStep::Start => Ok(vec![0]),
+            LedgerCursorStep::Balance => Ok(vec![1]),
+            LedgerCursorStep::Bytecode => Ok(vec![2]),
+            LedgerCursorStep::Datastore(key) => {
+                let mut bytes = vec![3];
+                if let Some(key) = key {
+                    bytes.extend(key.to_bytes());
+                }
+                Ok(bytes)
+            }
+            LedgerCursorStep::Finish => Ok(vec![4]),
```

Comment:
when is  `LedgerCursorStep::Finish` necessary ?

Reply:
I would have imagined something like:
when you ask for stuff strictly after AddressA Datastore(last_element), the cursor jumps to AddressB Balance.
Not sure it's the best solution in practice, just wondering.

Reply:
Yes I have think about it but when the ledger is finish what would be the cursor because it can't be AddressB Datastore(last_element) because then it will fetch again the last element of AddressB because the for loop include the key in the cursor. If I change it to not include the key then after the bytecode the cursor will not be Datastore(None) but Datastore(the first key) but on the client side it will be impossible to build this cursor before reading because he don't know the hash and then there will be a de-synchronization.